File: ros2_adapters/lio_benchmark_adapters/src/custommsg_to_pointcloud2.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <fstream>
#include <functional>
#include <iomanip>
#include <limits>
#include <memory>
#include <numeric>
#include <string>
#include <vector>

#include <livox_ros_driver2/msg/custom_msg.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>
#include <sensor_msgs/msg/point_field.hpp>

namespace {
constexpr uint32_t kPointStep = 22;

template <typename T>
void write_value(std::vector<uint8_t>& data, size_t offset, T value) {
  std::memcpy(data.data() + offset, &value, sizeof(T));
}

sensor_msgs::msg::PointField field(
    const std::string& name, uint32_t offset, uint8_t datatype) {
  sensor_msgs::msg::PointField result;
  result.name = name;
  result.offset = offset;
  result.datatype = datatype;
  result.count = 1;
  return result;
}
}  // namespace

class CustomMsgAdapter final : public rclcpp::Node {
 public:
  CustomMsgAdapter() : Node("lio_benchmark_custommsg_to_pointcloud2") {
    input_topic_ = declare_parameter<std::string>("input_topic", "/agt/sensors/lidar/custom");
    output_topic_ = declare_parameter<std::string>("output_topic", "/lio_benchmark/points");
    metrics_path_ = declare_parameter<std::string>("metrics_path", "");
    sort_by_time_ = declare_parameter<bool>("sort_by_time", true);
    auto input_qos = rclcpp::QoS(rclcpp::KeepLast(100)).best_effort().durability_volatile();
    publisher_ = create_publisher<sensor_msgs::msg::PointCloud2>(output_topic_, rclcpp::SensorDataQoS());
    subscription_ = create_subscription<livox_ros_driver2::msg::CustomMsg>(
        input_topic_, input_qos,
        std::bind(&CustomMsgAdapter::callback, this, std::placeholders::_1));
  }

  ~CustomMsgAdapter() override { save_metrics(); }

 private:
  void callback(const livox_ros_driver2::msg::CustomMsg::SharedPtr source) {
    ++frames_;
    input_points_ += source->points.size();
    for (size_t i = 1; i < source->points.size(); ++i) {
      input_time_backtracks_ += source->points[i].offset_time < source->points[i - 1].offset_time;
    }

    std::vector<size_t> selected;
    selected.reserve(source->points.size());
    for (size_t i = 0; i < source->points.size(); ++i) {
      const auto& point = source->points[i];
      const auto tag = static_cast<uint8_t>(point.tag & 0x30U);
      if (tag != 0x00U && tag != 0x10U) {
        ++invalid_tag_points_;
        continue;
      }
      if (!std::isfinite(point.x) || !std::isfinite(point.y) || !std::isfinite(point.z)) {
        ++non_finite_points_;
        continue;
      }
      selected.push_back(i);
    }
    if (sort_by_time_) {
      std::stable_sort(selected.begin(), selected.end(), [&](size_t lhs, size_t rhs) {
        return source->points[lhs].offset_time < source->points[rhs].offset_time;
      });
    }

    sensor_msgs::msg::PointCloud2 output;
    output.header = source->header;
    output.height = 1;
    output.width = static_cast<uint32_t>(selected.size());
    output.fields = {
        field("x", 0, sensor_msgs::msg::PointField::FLOAT32),
        field("y", 4, sensor_msgs::msg::PointField::FLOAT32),
        field("z", 8, sensor_msgs::msg::PointField::FLOAT32),
        field("intensity", 12, sensor_msgs::msg::PointField::FLOAT32),
        field("ring", 16, sensor_msgs::msg::PointField::UINT16),
        field("time", 18, sensor_msgs::msg::PointField::FLOAT32),
    };
    output.is_bigendian = false;
    output.point_step = kPointStep;
    output.row_step = output.point_step * output.width;
    output.is_dense = true;
    output.data.resize(output.row_step);

    uint32_t previous_offset = 0;
    bool first = true;
    for (size_t output_index = 0; output_index < selected.size(); ++output_index) {
      const auto& point = source->points[selected[output_index]];
      const size_t base = output_index * kPointStep;
      const float intensity = static_cast<float>(point.reflectivity);
      const uint16_t ring = point.line;
      const float time = static_cast<float>(point.offset_time) * 1.0e-9F;
      write_value(output.data, base + 0, point.x);
      write_value(output.data, base + 4, point.y);
      write_value(output.data, base + 8, point.z);
      write_value(output.data, base + 12, intensity);
      write_value(output.data, base + 16, ring);
      write_value(output.data, base + 18, time);
      output_time_backtracks_ += !first && point.offset_time < previous_offset;
      previous_offset = point.offset_time;
      first = false;
      ring_counts_.at(std::min<size_t>(ring, ring_counts_.size() - 1))++;
      time_min_s_ = std::min(time_min_s_, static_cast<double>(time));
      time_max_s_ = std::max(time_max_s_, static_cast<double>(time));
    }
    output_points_ += selected.size();
    publisher_->publish(std::move(output));
  }

  void save_metrics() const {
    if (metrics_path_.empty()) return;
    std::ofstream stream(metrics_path_);
    if (!stream) return;
    const auto dropped = input_points_ - output_points_;
    const double ratio = input_points_ ? static_cast<double>(dropped) / input_points_ : 0.0;
    stream << std::setprecision(15)
           << "{\n  \"implementation\": \"cpp\",\n"
           << "  \"frames\": " << frames_ << ",\n"
           << "  \"input_points\": " << input_points_ << ",\n"
           << "  \"output_points\": " << output_points_ << ",\n"
           << "  \"missing_fields\": 0,\n"
           << "  \"invalid_tag_points\": " << invalid_tag_points_ << ",\n"
           << "  \"non_finite_points\": " << non_finite_points_ << ",\n"
           << "  \"invalid_time_points\": 0,\n"
           << "  \"input_time_backtracks\": " << input_time_backtracks_ << ",\n"
           << "  \"output_time_backtracks\": " << output_time_backtracks_ << ",\n"
           << "  \"time_min_s\": " << (output_points_ ? time_min_s_ : 0.0) << ",\n"
           << "  \"time_max_s\": " << (output_points_ ? time_max_s_ : 0.0) << ",\n"
           << "  \"ring_min\": 0,\n  \"ring_max\": 3,\n"
           << "  \"ring_counts\": {\"0\": " << ring_counts_[0]
           << ", \"1\": " << ring_counts_[1] << ", \"2\": " << ring_counts_[2]
           << ", \"3\": " << ring_counts_[3] << "},\n"
           << "  \"dropped_points\": " << dropped << ",\n"
           << "  \"dropped_ratio\": " << ratio << ",\n"
           << "  \"time_semantics\": \"FLOAT32 seconds relative to CustomMsg header/timebase\",\n"
           << "  \"ring_semantics\": \"exact uint8 CustomPoint.line value widened to uint16; no synthetic rings\"\n}\n";
  }

  std::string input_topic_, output_topic_, metrics_path_;
  bool sort_by_time_ = true;
  rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr publisher_;
  rclcpp::Subscription<livox_ros_driver2::msg::CustomMsg>::SharedPtr subscription_;
  uint64_t frames_ = 0, input_points_ = 0, output_points_ = 0;
  uint64_t invalid_tag_points_ = 0, non_finite_points_ = 0;
  uint64_t input_time_backtracks_ = 0, output_time_backtracks_ = 0;
  std::array<uint64_t, 4> ring_counts_{};
  double time_min_s_ = std::numeric_limits<double>::infinity();
  double time_max_s_ = -std::numeric_limits<double>::infinity();
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CustomMsgAdapter>());
  rclcpp::shutdown();
  return 0;
}

File: evaluators/adapters/custommsg_to_pointcloud2.py
#!/usr/bin/env python3
"""Convert Livox CustomMsg to a common, timestamped PointCloud2 contract."""
from __future__ import annotations

import json
import math
import struct
from collections import Counter
from dataclasses import dataclass, field
from pathlib import Path
from typing import Iterable, Sequence

POINT_STRUCT = struct.Struct("<ffffHf")
FIELDS = (
    ("x", 0, 7), ("y", 4, 7), ("z", 8, 7), ("intensity", 12, 7),
    ("ring", 16, 4), ("time", 18, 7),
)


@dataclass
class Validation:
    frames: int = 0
    input_points: int = 0
    output_points: int = 0
    invalid_tag_points: int = 0
    non_finite_points: int = 0
    input_time_backtracks: int = 0
    output_time_backtracks: int = 0
    time_min_s: float | None = None
    time_max_s: float | None = None
    ring_counts: Counter[int] = field(default_factory=Counter)

    def as_dict(self) -> dict:
        dropped = self.input_points - self.output_points
        return {
            "frames": self.frames, "input_points": self.input_points, "output_points": self.output_points,
            "missing_fields": 0, "invalid_tag_points": self.invalid_tag_points,
            "non_finite_points": self.non_finite_points, "invalid_time_points": 0,
            "input_time_backtracks": self.input_time_backtracks, "output_time_backtracks": self.output_time_backtracks,
            "time_min_s": self.time_min_s, "time_max_s": self.time_max_s,
            "ring_min": min(self.ring_counts) if self.ring_counts else None,
            "ring_max": max(self.ring_counts) if self.ring_counts else None,
            "ring_counts": {str(k): v for k, v in sorted(self.ring_counts.items())},
            "dropped_points": dropped, "dropped_ratio": dropped / self.input_points if self.input_points else 0.0,
            "time_semantics": "FLOAT32 seconds relative to CustomMsg header/timebase",
            "ring_semantics": "exact uint8 CustomPoint.line value widened to uint16; no synthetic rings",
        }


def convert_points(points: Sequence, sort_by_time: bool = True, validation: Validation | None = None) -> bytes:
    stats = validation or Validation()
    stats.frames += 1
    stats.input_points += len(points)
    offsets = [int(point.offset_time) for point in points]
    stats.input_time_backtracks += sum(b < a for a, b in zip(offsets, offsets[1:]))
    selected = []
    for point in points:
        if (int(point.tag) & 0x30) not in (0x00, 0x10):
            stats.invalid_tag_points += 1
            continue
        xyz = (float(point.x), float(point.y), float(point.z))
        if not all(math.isfinite(value) for value in xyz):
            stats.non_finite_points += 1
            continue
        selected.append(point)
    if sort_by_time:
        selected.sort(key=lambda item: int(item.offset_time))
    buffer = bytearray(len(selected) * POINT_STRUCT.size)
    output_times: list[float] = []
    for index, point in enumerate(selected):
        seconds = int(point.offset_time) * 1e-9
        POINT_STRUCT.pack_into(buffer, index * POINT_STRUCT.size, float(point.x), float(point.y), float(point.z), float(point.reflectivity), int(point.line), seconds)
        output_times.append(seconds)
        stats.ring_counts[int(point.line)] += 1
    stats.output_points += len(selected)
    stats.output_time_backtracks += sum(b < a for a, b in zip(output_times, output_times[1:]))
    if output_times:
        stats.time_min_s = min(output_times) if stats.time_min_s is None else min(stats.time_min_s, min(output_times))
        stats.time_max_s = max(output_times) if stats.time_max_s is None else max(stats.time_max_s, max(output_times))
    return bytes(buffer)


def main() -> None:
    import rclpy
    from livox_ros_driver2.msg import CustomMsg
    from rclpy.node import Node
    from rclpy.qos import qos_profile_sensor_data
    from sensor_msgs.msg import PointCloud2, PointField

    class Converter(Node):
        def __init__(self) -> None:
            super().__init__("lio_benchmark_custommsg_to_pointcloud2")
            self.declare_parameter("input_topic", "/agt/sensors/lidar/custom")
            self.declare_parameter("output_topic", "/lio_benchmark/points")
            self.declare_parameter("sort_by_time", True)
            self.declare_parameter("metrics_path", "")
            self.stats = Validation()
            self.output_topic = str(self.get_parameter("output_topic").value)
            self.sort_by_time = bool(self.get_parameter("sort_by_time").value)
            self.metrics_path = str(self.get_parameter("metrics_path").value)
            self.pub = self.create_publisher(PointCloud2, self.output_topic, qos_profile_sensor_data)
            self.sub = self.create_subscription(CustomMsg, str(self.get_parameter("input_topic").value), self.callback, qos_profile_sensor_data)

        def callback(self, source: CustomMsg) -> None:
            output = PointCloud2()
            output.header = source.header
            output.height = 1
            output.width = 0
            output.is_bigendian = False
            output.is_dense = True
            output.fields = [PointField(name=n, offset=o, datatype=d, count=1) for n, o, d in FIELDS]
            output.point_step = POINT_STRUCT.size
            output.data = convert_points(source.points, self.sort_by_time, self.stats)
            output.width = len(output.data) // output.point_step
            output.row_step = output.width * output.point_step
            self.pub.publish(output)

        def save(self) -> None:
            if self.metrics_path:
                path = Path(self.metrics_path)
                path.parent.mkdir(parents=True, exist_ok=True)
                path.write_text(json.dumps(self.stats.as_dict(), ensure_ascii=False, indent=2) + "\n", encoding="utf-8")

    rclpy.init()
    node = Converter()
    try:
        rclpy.spin(node)
    finally:
        node.save(); node.destroy_node()
        if rclpy.ok(): rclpy.shutdown()


if __name__ == "__main__":
    main()
